### DISCUSSION TITLE
Surface precipitation

### DISCUSSION THREAD

**Discussion by jbarr444** [open-atmos/PySDM#1397] [General]
Hi @slayoo,

Can you explain how PySDM calculates surface precipitation in the 1D environment? I'm trying to split the precipitation into contributions from radius bins, but the total is ending up anywhere within one order of magnitude off from what the 'surface precipitation' product reports. I'm simply summing up 4/3 * pi * r**3 * multiplicity * v(r) / dz in the 0 height bin. v(r) is stokes drag and I'm assuming multiplicity has units # per unit area (?) I also tried summing the volume of particles that are less than height v(r)*dt, then dividing by dt, but that does similarly.

Thanks,
Jason

**Comment by slayoo** [comment]
Hi @jbarr444,

Multiplicity is unitless, it conveys the number of particles represented by a super droplet. We compute the surface flux only for the particles that cross the bottom boundary of the domain, not for all the particles in the bottom cell, the relevant code is here: https://github.com/open-atmos/PySDM/blob/163b6d9b805cacfd9adc02023d776bff8f42dae9/PySDM/backends/impl_numba/methods/displacement_methods.py#L176 Then, it is summed and normalised wrt area of the column base here: https://github.com/open-atmos/PySDM/blob/163b6d9b805cacfd9adc02023d776bff8f42dae9/PySDM/products/displacement/surface_precipitation.py#L34

As noted in #708, we are actually missing unit tests for the above code, help welcome!
HTH,
S.

**Reply by jbarr444** [reply]
Thanks, that is very helpful. I see this will be more complicated than I had hoped.

If you’re looking for someone to find ways to break your program, I might be able to help, but I don’t know if I’m qualified to fix them. On that topic, though, the unit is listed to default as m/s, but the formula involving the density of water implies it’s kg / m^2 / s ?

**Reply by slayoo** [reply]
Thank you for pointing the unit inconsistency!
Here's a PR addressing it and hopefully clarifying the code a bit: https://github.com/open-atmos/PySDM/pull/1399
